zoo_service.py

```diff
@@ -0,0 +1,107 @@
+from flask import Flask, request, abort
+from flask_sqlalchemy import SQLAlchemy
+from flask_marshmallow import Marshmallow
+from marshmallow import fields, ValidationError
+
+app = Flask(__name__)
+app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+mysqlconnector://myk_user:fwQH-totF8Pv@localhost:3306/lab6'
+db = SQLAlchemy(app)
+ma = Marshmallow(app)
+
+
+class Fish(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    weight_in_kg = db.Column(db.Float)
+    thermoregulation = db.Column(db.String(15))
+    lifetime_years = db.Column(db.Float)
+    animal_type = db.Column(db.String(50))
+    required_aquarium_capacity_liters = db.Column(db.Float)
+    required_temperature = db.Column(db.Integer)
+    required_lighting_level = db.Column(db.String(20))
+
+    def __init__(
+            self, weight_in_kg, thermoregulation, lifetime_years, animal_type,
+            required_aquarium_capacity_liters, required_temperature, required_lighting_level
+    ):
+        self.weight_in_kg = weight_in_kg
+        self.thermoregulation = thermoregulation
+        self.lifetime_years = lifetime_years
+        self.animal_type = animal_type
+        self.required_aquarium_capacity_liters = required_aquarium_capacity_liters
+        self.required_temperature = required_temperature
+        self.required_lighting_level = required_lighting_level
+
+
+class FishSchema(ma.Schema):
+    id = fields.Int()
+    weight_in_kg = fields.Float()
+    thermoregulation = fields.String()
+    lifetime_years = fields.Float()
+    animal_type = fields.String()
+    required_aquarium_capacity_liters = fields.Integer()
+    required_temperature = fields.Integer()
+    required_lighting_level = fields.String()
+
+
+fish_schema = FishSchema(exclude=["id"])
+fishes_schema = FishSchema(only=["id", "animal_type", "lifetime_years"], many=True)
+
+
+@app.route('/')
+def hello_page():
+    return 'Hello to the zoo aquariums!'
+
+
+@app.route('/fish', methods=['GET'])
+def get_all_fish():
+    all_fish = Fish.query.all()
+    return fishes_schema.jsonify(all_fish)
+
+
+@app.route('/fish/<id>', methods=['GET'])
+def get_fish(id):
+    fish = Fish.query.get(id)
+    if fish is None:
+        return abort(404)
+    return fish_schema.jsonify(fish)
+
+
+@app.route('/fish', methods=['POST'])
+def add_fish():
+    try:
+        params = fish_schema.load(request.json)
+        new_fish = Fish(**params)
+    except:
+        return abort(400)
```

Comment:
```suggestion
    try:
        params = fish_schema.load(request.json)
    except:
        return abort(400)
    new_fish = Fish(**params)
```

We only want to return HTTP 400 if the marshmallow validation has failed. Failure to create a `Fish` object can still result in a 500.

Reply:
An HTTP 400 is returned when the user made an incorrect request. Wouldn't it be better if failure to create **Fish** object will still return a 400 error, but in a separate try/except block and with an additional appropriate message?

Reply:
> Wouldn't it be better if failure to create Fish object will still return a 400 error, but in a separate try/except block and with an additional appropriate message?

No.

Marshmallow validation catches almost all possible cases of an invalid client request for you.
There is (almost) no way for an invalid request to pass Marshmallow validation but fail in `Fish(**params)`. Sending HTTP400 in a situation when there's a server failure (i.e., nothing's wrong with the client request) is misleading and incorrect.

You have to draw the line somewhere: foresee some of the situations, catch the exceptions that you can catch, and leave the rest be.

There's one more consideration here. Since almost no client request can lead to `Fish(**params)` failing, if it does fail then something has gone terribly wrong. In actual live products you almost never log invalid client requests, but unhandled exceptions and HTTP500s WILL get logged. Generally, you want a terriably wrong unforeseen situation to fail explicitly so that you're alerted that it happened.

Reply:
I changed it as you recommended. Thank you for the provided information.